### PR TITLE
Rosetta Code: S-Expressions

### PR DESCRIPTION
**Pull request by R4PaSs**
Simple S-Expressions parser, take only the two last commits into account, as the rest belongs to #1448.

### CODE REVIEW

**Comment by privat** [comment]
I'm not fan of having a common parser project. Each parser could be seen as a independent project in the picknit comprehensive archive.

**Comment by privat** [comment]
Oh... or maybe you wanted to provide a sequencer lib, thus closing the last checkbox of #466

**Comment by ventilooo** [comment]
+1

**Comment by R4PaSs** [comment]
The sequencer idea is good, I'll implement it for a future reroll of this PR and of #1448

**Comment by R4PaSs** [comment]
test this please

**Comment by R4PaSs** [comment]
Reroll, not depending on #1448 anymore, removed the parsers directory, `sexp.nit` is not a simple module in lib

File: lib/sexp.nit [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+class SExp
+	super SExpEntity
+
+	# Children of a SExp
+	var content = new Array[SExpEntity]
+
+	redef fun to_s do return "({content.join(" ")})"
+
+	# Returns a pretty-printable version of self
+	fun pretty_to_s: String do return recurse_to_s(0)
```
Comment (privat) [inline lib/sexp.nit]: example? (nitunit)

File: lib/sexp.nit [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+	# S-Exp compatible identifier
+	var content: String
+
+	redef fun to_s do return content
+end
+
+# An error parsing S-Expressions
+class SExpError
+	super SExpEntity
```
Comment (privat) [inline lib/sexp.nit]: super Error?

File: lib/sexp.nit [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	super StringProcessor
+
+	# Parses an S-Expression
+	fun parse_document: SExpEntity do
+		var stack = new Array[SExp]
+		var srclen = src.length
+		ignore_whitespaces
+		if pos >= srclen then return new SExpError("Empty S-expression")
+		var c = src[pos]
+		if c != '(' then return new SExpError("Invalid start of S-Expression, got `{c}`, exprected `(`.")
```
Comment (privat) [inline lib/sexp.nit]: an atom is a valid s-expression (by the inductive definition of s-expressions). You should not require to start with a `(`.

**Comment by R4PaSs** [comment]
Reroll, added Nitunit as @privat asked and corrected the parsing to support atoms and recursive S-Expressions

File: lib/sexp.nit (hunk 2 above)
Comment (R4PaSs) [inline lib/sexp.nit]: Cannot do because of issue #1552

**Comment by R4PaSs** [comment]
Reroll, tests are passing

**Comment by privat** [comment]
+1

**Pull request merged**